src/utils/encoders.ts

```diff
@@ -27,31 +27,82 @@ export const decodeFloat = (value: BigNumber) => {
   return value.mod(ONE).shl(value.div(ONE).toNumber());
 };
 
-export const encodeOrder = (order: DecodedOrder): EncodedOrder => {
+export const encodeOrders = ([order0, order1]: [DecodedOrder, DecodedOrder]): [
+  EncodedOrder,
+  EncodedOrder
+] => {
+  const liquidity0 = new Decimal(order0.liquidity);
+  const lowestRate0 = new Decimal(order0.lowestRate);
+  const highestRate0 = new Decimal(order0.highestRate);
+  const liquidity1 = new Decimal(order1.liquidity);
+  const lowestRate1 = new Decimal(order1.lowestRate);
+  const highestRate1 = new Decimal(order1.highestRate);
+
+  // if one order has 0 liquidity and the other has > 0 liquidity - use it to calculate z.
+  if (
+    liquidity0.eq(0) &&
+    liquidity1.gt(0) &&
+    lowestRate1.gt(0) &&
+    highestRate1.gt(0)
+  ) {
+    return [
+      encodeOrder(order0, calculateCorrelatedZ(order1)),
+      encodeOrder(order1),
+    ];
+  }
+  if (
+    liquidity1.eq(0) &&
+    liquidity0.gt(0) &&
+    lowestRate0.gt(0) &&
+    highestRate0.gt(0)
+  ) {
+    return [
+      encodeOrder(order0),
+      encodeOrder(order1, calculateCorrelatedZ(order0)),
+    ];
+  }
+  return [encodeOrder(order0), encodeOrder(order1)];
+};
+
+export const encodeOrder = (
+  order: DecodedOrder,
+  z?: BigNumber
+): EncodedOrder => {
   const liquidity = new Decimal(order.liquidity);
   const lowestRate = new Decimal(order.lowestRate);
   const highestRate = new Decimal(order.highestRate);
   const marginalRate = new Decimal(order.marginalRate);
+
   if (
     !(
       (highestRate.gte(marginalRate) && marginalRate.gt(lowestRate)) ||
-      (highestRate.eq(marginalRate) && marginalRate.eq(lowestRate))
+      (highestRate.eq(marginalRate) && marginalRate.eq(lowestRate)) ||
+      (highestRate.gt(marginalRate) &&
+        marginalRate.eq(lowestRate) &&
+        liquidity.isZero())
     )
-  ) {
+  )
     throw new Error(
       'Either one of the following must hold:\n' +
         '- highestRate >= marginalRate > lowestRate\n' +
         '- highestRate == marginalRate == lowestRate\n' +
+        '- (highestRate > marginalRate == lowestRate) AND liquidity == 0\n' +
         `(highestRate = ${highestRate}, marginalRate = ${marginalRate}, lowestRate = ${lowestRate})`
```

Comment:
Probably better to add `liquidity = ${liquidity}` here.

Reply:
done